### ISSUE TITLE
get-schema failed with the upgraded libyang version

### ISSUE THREAD

**Issue by navya46** [CESNET/Netopeer2#473]
Hi,
          With the previous yang revisions and libyang 0.16.133 and netopeer2-server 0.7.13,sysrepo version 0.7.7, get-schema was successfull.

Below is the output of sysrepoctl -l

[sysrepoctl_output_with_older_versions.txt](https://github.com/CESNET/Netopeer2/files/3728993/sysrepoctl_output_with_older_versions.txt)

Now for compatibility with other applications we have upgraded the libyang version to (1.1.52) and sysrepo version to 0.7.8 and  when i tried to extract get-schema for the modules it failed with the error "The requested schema was not found" and the below error was logged on netopeer2-server.
Oct 15 10:19:46 netopeer2-server[15569]: Invalid arguments (ly_ctx_get_module_by()).
Oct 15 10:19:46 netopeer2-server[15569]: Invalid arguments (ly_ctx_get_submodule()).
Oct 15 10:19:46 netopeer2-server[15569]: Session 6: thread 3 event new RPC.
Oct 15 10:19:46 netopeer2-server[15569]: Session 6: thread 3 event reply error.

Below is the output pf sysrepoctl -l
[sysrepoctl_output_with_latest_version.txt](https://github.com/CESNET/Netopeer2/files/3729018/sysrepoctl_output_with_latest_version.txt)

Thanks,
Navya

**Comment by michalvasko** [comment]
Hi Navya,
have you updated also *libnetconf2* and *netopeer2* to the latest release?

Regards,
Michal

**Comment by navya46** [comment]
Hi Michal,
We have upgraded libnetconf2 to version libnetconf2.so.0.12.51 till this commit we have taken  changes fa3c5aabef97a073b94da227e12e7526bff9f0cf  and  for netopeer2 till this commit 109cebaf262c56befbbc3108d21d302f44cba40b  we have considered changes.

Regards,
Navya

**Comment by michalvasko** [comment]
Hi Navya,
so you are not using the latest releases. Look, your error is a really weird one and since you are the only one having this problem, I must blame it on some incompatibility in your setup. So, my suggestion is to uninstall the old software (mainly run `make uninstall_with_repo` for *sysrepo*), then download the latest release of *libyang*, *sysrepo*, *libnetconf2*, and *netopeer2*, and install them. It should work then.

Regards,
Michal

**Comment by CC-vga** [comment]
Hello, 

It seems that I have a similar error, but when executing an edit-config. 
I'm using now the devel versions of libyang and libnetconf2, and the error message looks like this:
...
[ERR]: LY: Invalid arguments (lyd_schema_sort()).
[INF]: SR: Event "rpc" with ID 1 priority 0 succeeded.
Though, when I start the server again and make a get-config the updated values are there, so it rather seems that the previous error is wrongly thrown.
Could you suggest me please how to solve this?

All best
Voica

**Comment by michalvasko** [comment]
Hi,
you are right that it seems as some ignored error. Can you provide more information? Your `sysrepoctl -l` output, the edit-config you are applying and maybe even previous configuration is relevant.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi Michal, 

Now I realized that I get the error when I’m calling the get-config or lock and on server and the candidate data store contains what is now on cand0.txt (i.e., there is no configuration about IEEE 802.1Q bridge). Calling edit-config gives another error, but most probable this is due to the wrong flow.
Attached as models.txt is the output of sysrepoctl -l. If helps I added also the log of netopeer2-server when calling get-config. 

All best
Voica

[cand0.txt](https://github.com/CESNET/Netopeer2/files/4480300/cand0.txt)
[log.txt](https://github.com/CESNET/Netopeer2/files/4480301/log.txt)
[models.txt](https://github.com/CESNET/Netopeer2/files/4480302/models.txt)

**Comment by michalvasko** [comment]
Hi Voica,
thanks for all the details, I think the cause are your access rights (or lack of them) and when empty data are being passed where it is not expected, `NULL` is given as argument for `lyd_schema_sort()`. But, because of this I am not able to reproduce it and find the exact spot. I believe the error is harmless but if you were able to get me a stack trace of the call that fails, I should be able to fix it.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi Michal,

What you mean by "stack trace of the call"? I suppose that the log of netopeer2-server is not enough. So how can I get the log of libyang?

Thank you
Voica
________________________________

**Comment by michalvasko** [comment]
Hi,
I meant using a debugger (for example `gdb`), break execution at that function when it fails, and then print the current stack (command `bt` in `gdb`). If it is not too much trouble.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi,

It seems a bit problematic. I'll see what I'm up to!

All best
Voica
________________________________

**Comment by CC-vga** [comment]
Hi again,



Now I realized that is rather impossible to get a stack trace; having no overview of the code I don’t know where to put break points, where to step in or go next. There is no other alternative to fix this problem?



Thank you

Voica

________________________________

**Comment by michalvasko** [comment]
Hi,
not really if I cannot reproduce it. But setting a breakpoint should be easy. I believe the error printed is defined [here](https://github.com/CESNET/libyang/blob/devel/src/tree_data.c#L4844) so if you set the breakpoint there and print the stack trace once the execution stops there, it may be all that I need.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi, 

It is a bit strange, but after a system update, I cannot reproduce the error anymore. Now on consequent calls on get-config I get the error 
[ERR]: SR: Failed to truncate shared memory (Invalid argument).
but this is another problem that I might solve by myself. I would like to, but unfortunately git doesn't allow me to close this issue.
Sorry for bothering!

All best
Voica

**Comment by michalvasko** [comment]
Hi,
can I ask on what platform are you exactly?

Regards,
Michal

**Comment by CC-vga** [comment]
Hi, 

Sure, the netopeer server is on Ubuntu 18.04, under Windows sub-system for Linux. If I can assist you with other information please don't hesitate!

All best
Voica

**Comment by michalvasko** [comment]
Hi Voica,
yeah, that is what I thought when you posted the truncate error. So, for future reference, please always mention what platform are you on unless it is a native (or virtual) x68_64 Linux distribution. We certainly do not support the Windows subsystem because you never know what will not work there. I would strongly suggest you switch to a virtual Linux instead.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi Michal, 

Thank you very much, it seems that you saved me from a lot of frustration! But there is any explicit note on netopeer or libyang page that WSL is not fully supported?

All best
Voica

**Comment by michalvasko** [comment]
Hi,
let me rephrase that, Windows Linux subsystem should have an explicit note that not all Linux features are supported. But, to be honest, I assumed that the first time I heard about it.

Regards,
Michal

**Comment by CC-vga** [comment]
Hi,

Yes, that's also true!

Good luck with that
Voica
________________________________
